andrepereiradasilva/joomla-cms
# try to solve remaining acl problems - initial commit

Pull Request for Several ACL Issues (part 2).
#### Summary of Changes

What this PR does:
- Further simplifies the ACL rules code
- Displays the correct calculated permissions in article <- category <- com_content <- global config scenario
- Stop explicit writing the parent asset permissions on creating a new item (ex: module)
- Solves some other minor bugs
  ...
#### Testing Instructions
##### Pre-requisites

Use latest staging.
Clean install.
##### Tests

1: This tests Global config child groups Inherit
- Go to Global config
- Change Public "Delete" to Denied
- Change Public "Edit" to Allowed
- Refresh the page.
- Change Registered "Edit state" to Denied
- Change Registered "Edit Own" to Allowed
- Refresh the page
- Check all Global config Permissions are all inhrited as they should.

2: This tests Global config -> Component Inherit
- Go to Articles (com_content) Config 
- Check the com_config permission are all inhrited as they should.
- Change Author "Edit Own" to Denied
- Refresh the page
- Check the child groups permissions are all inhrited as they should.

3: This tests Global config -> Component -> Item Inherit
- Create new category, name it "Test ACL" don't change anything and save
- Now edit the category
- Check the com_config + com_content permissions are all inhrited as they should.
- Change Editor "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

4: This tests Global config -> Component -> Extension -> Item Inherit
- Create a new article assign to "Test ACL" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Change Manager "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

5: This tests creating anew item without explicit inhreting permissions
- Create a new article assign to "Test ACL 2" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Check there is no explict "Allowed" or "Denied" permission.
#### Know issues for other(s) PR

1: If a component as a permission that doesn't exists in global config (ex: frontend editing in com_modules) we get "Not Allowed (Inherited)" when we should get "Not Allowed (Default)"

2: When creating a new item (not saving) if uses the calculated permissions from the component (ex: module <- com_modules <- global config), but not if you have a section too (ex: article <- category <- com_content <- global config)

3: In"2:", it uses the component permission, but should use the child of componet calculated permissions.

4: When changing a permission of an item that doesn't have a row in the asset table the row is created fine for module <- com_modules <- global config scenario, but not for article <- category <- com_content <- global config scenario.
